sdk/python/lib/pulumi/output.py

```diff
@@ -171,7 +171,7 @@ def apply(self, func: Callable[[T], Input[U]], run_with_unknowns: Optional[bool]
                     # contain any unknown values, collapse its value to the unknown value. This ensures that callbacks
                     # that expect to see unknowns during preview in outputs that are not known will always do so.
                     if not is_known and run_with_unknowns and not contains_unknowns(value):
-                        value = UNKNOWN
+                        value = cast(T, UNKNOWN)
```

Comment:
Should we be somehow encoding `UNKNOWN` in the type system here?  Really not sure about that.  It seems it could potentially help us catch a whole other class of subtle issues.  cc @pgavlin @CyrusNajmabadi .

Reply:
agreed.  i don't have a great answer here beyond "there's an internal type we have that encodes all of this" and "there's a public type which hides this".